expression/builtin_arithmetic.go

```diff
@@ -0,0 +1,270 @@
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/util/types"
+)
+
+var (
+	_ functionClass = &arithmeticPlusFunctionClass{}
+	_ functionClass = &arithmeticFunctionClass{}
+)
+
+var (
+	_ builtinFunc = &builtinArithmeticPlusRealSig{}
+	_ builtinFunc = &builtinArithmeticPlusIntSig{}
+	_ builtinFunc = &builtinArithmeticPlusIntUnsignedSig{}
+	_ builtinFunc = &builtinArithmeticPlusDecimalSig{}
+	_ builtinFunc = &builtinArithmeticSig{}
+)
+
+func numericContextResultType(ft *types.FieldType) types.TypeClass {
+	if types.IsTemporalType(ft.Tp) {
+		if ft.Decimal > 0 {
+			return types.ClassDecimal
+		}
+		return types.ClassInt
+	}
+	if ft.ToClass() == types.ClassString {
+		return types.ClassReal
+	}
+	return ft.ToClass()
+}
+
+type arithmeticPlusFunctionClass struct {
+	baseFunctionClass
+}
+
+// setFlenDecimal4Int is called to set proper `Flen` and `Decimal` of return
+// type according to the two input parameter's types.
+func (c *arithmeticPlusFunctionClass) setFlenDecimal4Int(retTp, a, b *types.FieldType) {
+	retTp.Decimal = 0
+	if a.Flen == types.UnspecifiedLength || b.Flen == types.UnspecifiedLength {
+		retTp.Flen = mysql.MaxIntWidth
+	} else {
+		retTp.Flen = int(math.Max(float64(a.Flen), float64(b.Flen))) + 1
+		retTp.Flen = int(math.Min(float64(retTp.Flen), float64(mysql.MaxIntWidth)))
+	}
+}
+
+// setFlenDecimal4Real is called to set proper `Flen` and `Decimal` of return
+// type according to the two input parameter's types.
+func (c *arithmeticPlusFunctionClass) setFlenDecimal4RealOrDecimal(retTp, a, b *types.FieldType, isReal bool) {
+	if a.Decimal != types.UnspecifiedLength && b.Decimal != types.UnspecifiedLength {
+		retTp.Decimal = int(math.Max(float64(a.Decimal), float64(b.Decimal)))
+		if a.Flen == types.UnspecifiedLength || b.Flen == types.UnspecifiedLength {
+			retTp.Flen = types.UnspecifiedLength
+		} else {
+			digitsInt := int(math.Max(float64(a.Flen-a.Decimal-1), float64(b.Flen-b.Decimal-1)))
+			retTp.Flen = digitsInt + retTp.Decimal + 3
+			if isReal {
+				retTp.Flen = int(math.Min(float64(retTp.Flen), float64(mysql.MaxRealWidth)))
+			} else {
+				retTp.Flen = int(math.Min(float64(retTp.Flen), float64(mysql.MaxDecimalWidth)))
+			}
+		}
+	} else {
+		retTp.Decimal = types.UnspecifiedLength
+		retTp.Flen = types.UnspecifiedLength
+	}
+}
+
+func (c *arithmeticPlusFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	tpA, tpB := args[0].GetType(), args[1].GetType()
+	tcA, tcB := numericContextResultType(tpA), numericContextResultType(tpB)
+	if tcA == types.ClassReal || tcB == types.ClassReal {
+		bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpReal, tpReal, tpReal)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if mysql.HasUnsignedFlag(tpA.Flag) || mysql.HasUnsignedFlag(tpB.Flag) {
```

Comment:
no, result type of function plus with two float point parameters is always unsigned